### PR TITLE
Remaining CssValueInput todos

### PR DESCRIPTION
**Pull request by ghost**
Solves the issues presented in: https://github.com/webstudio-is/webstudio-designer/issues/333

## Before requesting a review

- [ ] if the PR is WIP - use draft mode
- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")
- [ ] what kind of review is needed?
  - [x] conceptual
  - [ ] detailed
  - [ ] with testing

## Test cases

- [ ] step by step interaction description and what is expected to happen

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio-designer/blob/main/apps/designer/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `designer/env-check.js` if mandatory

### CODE REVIEW

File: apps/designer/app/designer/features/style-panel/shared/css-value-input/css-value-input.tsx
Comment (kof) [inline apps/designer/app/designer/features/style-panel/shared/css-value-input/css-value-input.tsx:273]: what is auto solving here?

File: apps/designer/app/designer/features/style-panel/shared/css-value-input/unit-select.tsx
Comment (kof) [inline apps/designer/app/designer/features/style-panel/shared/css-value-input/unit-select.tsx:55]: this can use toValue, right?

File: packages/design-system/src/components/combobox.tsx
Comment (kof) [inline packages/design-system/src/components/combobox.tsx:27]: nit: `py: '$1' `reads better

File: packages/react-sdk/src/css/units.ts
Comment (kof) [inline packages/react-sdk/src/css/units.ts:32]: file was generated, its even in the comment above

File: packages/react-sdk/src/stitches/to-css.ts
Comment (kof) [inline packages/react-sdk/src/stitches/to-css.ts:5]:
```suggestion
export const toValue = (value?: StyleValue): string => {
```

File: packages/react-sdk/src/stitches/to-css.ts
Comment (kof) [inline packages/react-sdk/src/stitches/to-css.ts:8]: are you saying that we now have a unit type that has an empty string value.unit instead of number?

File: packages/react-sdk/src/css/units.ts
Comment (kof) [reply packages/react-sdk/src/css/units.ts:32]: I added export const Unit = z.union([z.enum(units), z.literal("number")]); in css/schema number as a valid one, but that's kinda wrong, these two are now out of sync

File: apps/designer/app/designer/features/style-panel/shared/css-value-input/css-value-input.tsx
Comment (kof) [inline apps/designer/app/designer/features/style-panel/shared/css-value-input/css-value-input.tsx:114]: there are 2 problems here:
1. you are accessing parent element in a way that is fragile - if we change the structure - this will suddenly start adding attribute on a wrong element
3. focus has actual meaning in the sense it is focused, but in this case you are actually unfocusing, so you are kinda trying to do something like visually-focused or something. 

I just needed the same thing [here](https://github.com/webstudio-is/webstudio-designer/pull/351/files#diff-e2bb0fb4fd1b7d2d36fb8c2c21769648fc2154d3226fae44d2e559357d648639R176)

 and called it. "active" like the active CSS pseudo selector ... I am not sure if my semantics better, wdyt?
 also I did it over prop, so to set it you would have to set it over local state

Comment (kof) [reply apps/designer/app/designer/features/style-panel/shared/css-value-input/css-value-input.tsx:114]: I should probably turn it into variant="active", because its just a visual thing and `variant="ghost" active` shouldn't be possible at the same time I guess

Comment (kof) [reply apps/designer/app/designer/features/style-panel/shared/css-value-input/css-value-input.tsx:114]: changed it to be variant over there

File: apps/designer/app/designer/features/style-panel/shared/css-value-input/unit-select.tsx
Comment (kof) [reply apps/designer/app/designer/features/style-panel/shared/css-value-input/unit-select.tsx:55]: turns out same thing has been already used in `parseCssValue` function which should also use toValue ... we need a single place that turns internal CSS value object to CSS string

**Comment by kof** [comment]
- [ ] Some values can't have a numeric value, e.g. overflow: 3px would be illegal

We already have that information from the CSS database so we could prevent even going into that mode or alternatively fix validation of the input. Right now it allows to enter unit value for overflow for example for some reason. It should be in an invalid state with that value

File: apps/designer/app/designer/features/style-panel/shared/css-value-input/css-value-input.tsx
Comment (kof) [reply apps/designer/app/designer/features/style-panel/shared/css-value-input/css-value-input.tsx:114]: will merge totay-tomorrow that other pr so you can use the variant

**Comment by taylornowotny** [comment]
This is looking great! The click-and-hold to scrub interaction is really smooth. I think users are going to like that

File: apps/designer/app/designer/features/style-panel/shared/css-value-input/css-value-input.tsx
Comment (kof) [reply apps/designer/app/designer/features/style-panel/shared/css-value-input/css-value-input.tsx:273]: I found a solution for all known icon inside textfield problems: check this out, lets wait until I merge and we can rebase this one ad4e57a83e28bea3538edd057e9fef9c724dfb7e

Comment (kof) [reply apps/designer/app/designer/features/style-panel/shared/css-value-input/css-value-input.tsx:273]: should be able to remove width:auto with the updates in main

File: apps/designer/app/designer/features/style-panel/shared/css-value-input/css-value-input.tsx
Comment (kof) [reply apps/designer/app/designer/features/style-panel/shared/css-value-input/css-value-input.tsx:114]: btw this has been merged, you can set that state as a prop state="active"

File: apps/designer/app/designer/features/style-panel/shared/css-value-input/css-value-input.tsx
Comment (kof) [inline apps/designer/app/designer/features/style-panel/shared/css-value-input/css-value-input.tsx:89]: its a boolean, so it can be named isInputActive

File: apps/designer/app/designer/features/style-panel/shared/get-final-value.ts
Comment (kof) [inline apps/designer/app/designer/features/style-panel/shared/get-final-value.ts:35]: same here with assertion

File: apps/designer/app/designer/features/style-panel/shared/parse-css-value.ts
Comment (kof) [inline apps/designer/app/designer/features/style-panel/shared/parse-css-value.ts:6]: same here with assertions

File: packages/design-system/src/components/panel.tsx
Comment (kof) [inline packages/design-system/src/components/panel.tsx:7]: I remember having updated this shadow from figma, just a few days ago, lets 
1. discuss how this is wrong
2. normalize asking questions if you see we are back and forth on something

File: packages/design-system/src/components/primitives/numeric-gesture-control.ts
Comment (kof) [inline packages/design-system/src/components/primitives/numeric-gesture-control.ts:42]: why would that be unknown

File: packages/design-system/src/components/primitives/numeric-gesture-control.ts
Comment (kof) [inline packages/design-system/src/components/primitives/numeric-gesture-control.ts:42]: thats timerId

Comment (kof) [reply packages/design-system/src/components/primitives/numeric-gesture-control.ts:42]: timeout is the argument you pass to setTimeout, what you receive is an id for that timer

File: apps/designer/app/designer/features/style-panel/shared/css-value-input/unit-select.tsx
Comment (Andarist) [reply apps/designer/app/designer/features/style-panel/shared/css-value-input/unit-select.tsx:59]: the object passed to `toValue` has a property `value` that is of type: `number | undefined` but the expected type of that property is `string | number | string[]`. Here, `undefined` is not compatible with the expected type. The `undefined` appear here because the `value` identifier here has an optional value (its `value` property is `number` though, so it's compatible).

So for example this works OK:
```ts
          if (value) {
            toValue({ ...value, unit });
          }
```
But we need to also handle the other case somehow (when `value === undefined`). 

This one type checks OK:
```ts
  const renderUnits = useMemo(
    () =>
      value &&
      units
        .filter((unit) => {
          return isValid(property, toValue({ ...value, unit }));
        })
        .map((unit) => unitRenderMap.get(unit) ?? unit),
    [units, property, value]
  );
```

but it changes the type of `renderUnits` from `string[]` to `string[] | undefined` and this causes a problem later on.

In practice, this won't be a problem because we have an early return based on `value == undefined` but we can't express a relationship between `value` and `renderUnits` in a way that would allow TS to understand that checking `value` also checks~ `renderUnits` (unless we combine both into an object/array somehow).

The easiest fix is to change the latter check. So the next version could look smth like this:
```ts
  const renderUnits = useMemo(
    () =>
      value &&
      units
        .filter((unit) => {
          return isValid(property, toValue({ ...value, unit }));
        })
        .map((unit) => unitRenderMap.get(unit) ?? unit),
    [units, property, value]
  );

  const renderValue = value && unitRenderMap.get(value.unit);

  if (renderUnits == undefined || renderValue === undefined) {
    return [isOpen, null];
  }
```
That, in turn, will result in some further problems with the argument passed to `onChange`. So it seems that the easiest fix is to just check all of those things in the `if` responsible for the early return. The final version could look smth like this:
```ts
  const renderUnits = useMemo(
    () =>
      value &&
      units
        .filter((unit) => {
          return isValid(property, toValue({ ...value, unit }));
        })
        .map((unit) => unitRenderMap.get(unit) ?? unit),
    [units, property, value]
  );

  const renderValue = value && unitRenderMap.get(value.unit);

  if (value === undefined || renderUnits == undefined || renderValue === undefined) {
    return [isOpen, null];
  }
```

Is this extra added check somewhat redundant? Yes. Is this extra added check purely to satisfy the TS? Yes. 😉

File: apps/designer/app/designer/features/style-panel/shared/get-final-value.ts
Comment (Andarist) [reply apps/designer/app/designer/features/style-panel/shared/get-final-value.ts:35]: The reported error here is just terrible:
> Type '"invalid"' is not assignable to type '"unit"'.ts(2322)

The presence of the error is legit though and by casting here we risk some runtime failures elsewhere. You want to end up with `InvalidValue` here. That type has its `value` property typed as `string`:
https://github.com/webstudio-is/webstudio-designer/blob/cfd40d748b936d23caa3b6c60a8385c5df7f77c8/packages/react-sdk/src/css/schema.ts#L34
but the `currentValue`'s type here is `UnitValue` (we are within `if (currentValue.type === "unit")` here) and its `value` property is typed as `number`:
https://github.com/webstudio-is/webstudio-designer/blob/cfd40d748b936d23caa3b6c60a8385c5df7f77c8/packages/react-sdk/src/css/schema.ts#L12

File: apps/designer/app/designer/features/style-panel/shared/parse-css-value.ts
Comment (Andarist) [reply apps/designer/app/designer/features/style-panel/shared/parse-css-value.ts:6]: `.concat` is a generic function and its type definitions only allow you to pass items of the type contained in the source array. Basically, for a given `T[]` concat can only produce `T[]`. Your `T` here is a list of some string literals (representing units) and `"number"` is not one of them thus it's not accepted as a valid argument because it's not assignable to `T`.

As a workaround for this I recommend using spread - which allows you to create new array types out of the input "arguments":
```ts
const unitRegex = new RegExp(`${[...units, "number"].join("|")}`);
```

File: packages/design-system/src/components/panel.tsx
Comment (kof) [reply packages/design-system/src/components/panel.tsx:7]: We seem to be looking at different figma, lets figure this out:

## case 1: 

https://www.figma.com/file/EfzZ2w94KpmUviZJSBgxCd/Designer-UI?node-id=1602%3A41906

```
box-shadow: 0px 2px 7px rgba(0, 0, 0, 0.1), 0px 5px 17px rgba(0, 0, 0, 0.15);
```

Screenshot 2022-11-01 at 10 37 39


## case 2

https://www.figma.com/file/EfzZ2w94KpmUviZJSBgxCd/Designer-UI?node-id=2327%3A47295

```
box-shadow: 0px 2px 7px rgba(0, 0, 0, 0.1), 0px 5px 17px rgba(0, 0, 0, 0.15);
```

Screenshot 2022-11-01 at 10 40 11

Comment (kof) [reply packages/design-system/src/components/panel.tsx:7]: @natlusyht is @taylornowotny helpful? I think he confirmed what I was saying

**Review by kof — CHANGES_REQUESTED** [review]
lst thing is with the shadow

File: packages/design-system/src/components/panel.tsx
Comment (kof) [reply packages/design-system/src/components/panel.tsx:7]: ohhh now I understand

Comment (kof) [reply packages/design-system/src/components/panel.tsx:7]: btw background color is gray4 instead of slate4, it was already wrong, not a regression, but since we looked please take it along

**Review by kof — APPROVED** [review]

**Pull request merged**
